### PR TITLE
Pyflyte build imageSpec

### PR DESCRIPTION
**Pull request by pingsutw**
# TL;DR
[RFC](https://docs.google.com/document/d/1cS1ZlilJisGTYTtmaG_zNjFRcuzUGZLP1n3oHEhrZ9o/edit#)
We add a new arg `imageSpec`, user can specify it in the either `@task` or `@workflow`, then flytekit will build the image at registration time.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [x] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Example:
```python
from flytekit import task, workflow
from flytekit.image_spec.image_spec import ImageSpec


@task
def t2() -> str:
    print("hello")
    return "hello"


@task(image_spec=ImageSpec(packages=["pandas", "numpy"], registry="pingsutw"))
def t1() -> str:
    print("hello")
    print("hello")
    return "hello"


@workflow(image_spec=ImageSpec(
    packages=["git+https://github.com/flyteorg/flytekit@4834ba03470b66f2f7c800c6b08e77062f2442b6"],
    apt_packages=["git"],
    registry="pingsutw")
)
def wf():
    t1()
    t2()


if __name__ == '__main__':
    wf()

```

image

image

image.lock:
```json
{
  "pingsutw": [
    "9zGai7ij01MWkewVim1cxg..",
    "2mvWPRSd8LMbtyhyanIO5Q..",
    "48b32eb8bbd164cbcf021f4c944c3aed",
    "b91e0d14186edfef00d7ffafd642ffa2",
    "dba9147b996c9d40a8a612fb34dc5ce7",
    "850d8a35b44ba2694019256e560d2770",
    "dff95b42bf5c3273942dd8cf2f7dbbaa",
    "613633db1f727f3645621a5b7b9e9b9f",
    "05db5f898e1a5ccdc2796ebea33a4454"
  ]
}
```

## Tracking Issue
https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
TODO:
1.  Docs
2. pyflyte build --remote

### CODE REVIEW

File: flytekit/extend/image_spec/base_image.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        packages: list of packages that will be installed in the image.
+        os: operating system. by default is ubuntu 20.04.
+        registry: docker registry. if it's specified, flytekit will push the image.
+        python_version: python version in the image.
+    """
+
+    packages: List[str]
+    base_image: str = "pingsutw/envd_base:v8"
+    registry: Optional[str] = None
+    python_version: Optional[str] = "3.9"
```
Comment (wild-endeavor) [inline flytekit/extend/image_spec/base_image.py:29]: let's make this default to the currently running version of python.

File: flytekit/extend/__init__.py [hunk 2]
```diff
@@ -42,6 +42,4 @@
 from flytekit.core.promise import Promise
 from flytekit.core.python_customized_container_task import PythonCustomizedContainerTask
 from flytekit.core.shim_task import ExecutableTemplateShimTask, ShimTaskExecutor
-from flytekit.core.task import TaskPlugins
 from flytekit.core.type_engine import DictTransformer, T, TypeEngine, TypeTransformer
-from flytekit.tools.translator import get_serializable
```
[the remark this answers is not in the document]
Comment (wild-endeavor) [reply flytekit/extend/__init__.py:47]: is this a circular dependency issue?  let's move the image spec stuff out of extend.

File: flytekit/extend/image_spec/base_image.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    for pkg in image_spec.packages:
+        packages_list += f'"{pkg}", '
+
+    envd_config = f"""# syntax=v1
+
+def build():
+    base(image="{image_spec.base_image}", dev=False)
+    install.python_packages(name = [{packages_list}])
+    install.python(version="{image_spec.python_version}")
+    runtime.environ(env={{"PYTHONPATH": "/"}})
```
Comment (wild-endeavor) [inline flytekit/extend/image_spec/base_image.py:43]: why does this need to be /?

File: flytekit/tools/translator.py
Comment (wild-endeavor) [inline flytekit/tools/translator.py:1]: i feel like this file shouldn't be changing... can we move it out?

**Review by wild-endeavor — COMMENTED** [review]
Can you walk me through what happens in the following scenario if we `pyflyte run`?  What about `pyflyte register --non-fast`?

```python
@task(image_spec=ImageSpec(packages=["colorama==0.2.1"], base_image="ghcr.io/flyteorg/flytekit:py3.10-1.4.2"))
def say_hello() -> str:
    return "hello world"


@dynamic
def dt1() -> str:
    return say_hello()


@dynamic
def dt2() -> str:
    from other_lib import say_bonjour, say_cava
    tmp = say_cava()
    return say_bonjour(i=tmp)


@workflow
def my_wf_1() -> str:
    return dt1()


@workflow
def my_wf_2() -> str:
    return dt2()


# other_lib.py
@task(image_spec=ImageSpec(packages=["colorama==0.2.1"], base_image="ghcr.io/flyteorg/flytekit:py3.10-1.4.2"))
def say_bonjour(i: str) -> str:
    ...

@task(image_spec=ImageSpec(packages=["other_package==0.2.1"], base_image="ghcr.io/flyteorg/flytekit:py3.10-1.4.2"))
def say_cava() -> str:
    ...

```

What are the values of the `_F_SS_C` env vars?

File: flytekit/remote/remote.py [hunk 4]
```diff
@@ -548,8 +548,6 @@ def raw_register(
             return None
 
         if isinstance(cp_entity, task_models.TaskSpec):
-            if isinstance(cp_entity, FlyteTask):
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/remote/remote.py:551]: The type of `cp_entity` is always `TaskSpec`, so this line will never be true. Not sure why we add this here

File: flytekit/extend/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/extend/__init__.py:47]: reverted it

File: flytekit/extend/image_spec/base_image.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+    return cfg_path
+
+
+def build_docker_image(image_spec: ImageSpec, name: str, tag: str):
+    if should_build_image(image_spec.registry, tag) is False:
+        return
+
+    cfg_path = create_envd_config(image_spec)
+    print("building image...")
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/extend/image_spec/base_image.py]: updated it

File: flytekit/tools/translator.py [hunk 6]
```diff
@@ -155,6 +158,19 @@ def fn(settings: SerializationSettings) -> List[str]:
     return fn
 
 
+def _update_entity_image(settings: SerializationSettings, entity: FlyteLocalEntity):
+    if not isinstance(entity, (PythonAutoContainerTask, WorkflowBase)) or not entity.image_spec:
+        return
+    if settings.fast_serialization_settings.enabled:
+        tag = calculate_hash_from_image_spec(entity.image_spec)
+        settings.fast_serialization_settings.destination_dir = "/"
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/tools/translator.py]: yup, added

File: flytekit/extend/image_spec/base_image.py (hunk 1 above)
Comment (pingsutw) [reply flytekit/extend/image_spec/base_image.py]: done

File: flytekit/extend/image_spec/base_image.py (hunk 3 above)
Comment (pingsutw) [reply flytekit/extend/image_spec/base_image.py]: removed it

**Review by fg91 — COMMENTED** [review]
This is a very nice feature. At my current company we build all images in CI so we won't use this much (apart from quick experimentation) but at my previous company we built our own cli for building images on the fly during workflow registration (in the form of a [hydra](https://hydra.cc/) launcher, see [talk](https://youtu.be/tghvVvHJi7s?t=1034) at the oss sync). So I absolutely see a need for this feature and it is great that this is being tackled 🚀 :)

If I understand it correctly, there will be two ways to specify an image definition:

1. Via envd
2. By providing a base image and a list of pip and apt packages etc.

However, for me personally and most engineers I know, the most useful way would be to simply provide the path to a Dockerfile in the image spec:

```python
ImageSpec(
    registry="gcr.io/...",
    file="some/path/Dockerfile",
)
```

envd looks very promising (I didn't know it so far), e.g. in terms of build time. Option 2 is definitely nice for data scientists due to its ease of use. But Dockerfiles are widely adopted and most engineers know how to write them. The [Spark Dockerfile in the documentation](https://docs.flyte.org/projects/cookbook/en/latest/auto/integrations/kubernetes/k8s_spark/index.html#how-to-build-your-dockerfile-for-spark-on-kubernetes), for instance, couldn't be built with the current image spec.

What do you think about having an abstract `ImageSpec` base class without the base image and pip packages attributes, ... and then, in addition to the current options, provide one for a Dockerfile as well?

File: flytekit/core/task.py [hunk 7]
```diff
@@ -84,7 +85,8 @@ def task(
     interruptible: Optional[bool] = None,
     deprecated: str = "",
     timeout: Union[_datetime.timedelta, int] = 0,
-    container_image: Optional[str] = None,
+    container_image: Optional[Union[str, ImageSpec]] = None,
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/core/task.py:88]: we'll delete image_spec arg in the task

File: flytekit/image_spec/image_spec.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+@dataclass_json
+@dataclass
+class ImageSpec:
+    """
+    This class is used to specify the docker image that will be used to run the task.
+
+    Args:
+        registry: docker registry. if it's specified, flytekit will push the image.
+        packages: list of python packages that will be installed in the image.
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/image_spec/image_spec.py:23]: yes, we will add name to imageSpec

**Comment by pingsutw** [comment]
@fg91 Thanks for the great feedback and review. we had an internal discussion yesterday, I'm going to refactor this PR.

> What do you think about having an abstract ImageSpec base class without the base image and pip packages attributes, ... and then, in addition to the current options, provide one for a Dockerfile as well?

That's why we want put it to the plugin. some of users like to use dockerfile, some folk don't. we can have another plugin (`flytekitplugins-docker`), it can build the docker image if dockerfile is specified.

The pyflyte run behaviour will be like:
```
pyflyte run --build_engine envd/docker wf.py ...
```

File: flytekit/image_spec/image_spec.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    # replace "=" with "." to make it a valid tag
+    return tag.replace("=", ".")
+
+
+def image_exist(registry: str, tag: str) -> bool:
+    """
+    Check if the image has been pushed to the registry.
+    """
+    if os.path.isfile(IMAGE_LOCK) is False:
+        return True
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/image_spec/image_spec.py:70]: we decide to use docker sdk to check if the image exists in the registry. will remove lock file

File: plugins/flytekit-envd/flytekitplugins/envd/__init__.py [hunk 10]
```diff
@@ -0,0 +1,13 @@
+"""
+.. currentmodule:: flytekitplugins.envd
+
+This plugin enables seamless integration between Flyte and mlflow.
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply plugins/flytekit-envd/flytekitplugins/envd/__init__.py:4]: fixed, thanks

File: flytekit/image_spec/image_spec.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (fg91) [reply flytekit/image_spec/image_spec.py:70]: I think this is a good idea. I was very unsure whether I like the lockfile mechanism when reading it.

How will authentication work when using the python docker sdk? Is there a way to check the existence of a private image, without pulling it, in a way that the python client picks up the creds from `docker login`? Or does flytekit need to store credentials? I'm asking because in our internal registration helper script we have a function that retrieves the manifest of an image from the registry (with `requests.get`) in order to calculate the image digest. This helper requires an oauth token which is generated using a gcp auth library. This of course doesn't generalize well to other providers...

File: flytekit/core/task.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (fg91) [reply flytekit/core/task.py:88]: Ok, cool, this is nice.

**Comment by fg91** [comment]
> we can have another plugin (`flytekitplugins-docker`), it can build the docker image if dockerfile is specified.

Ok awesome!

> The pyflyte run behaviour will be like:
> 
> ```
> pyflyte run --build_engine envd/docker wf.py ...
> ```

Is specifying the `--build_engine` actually necessary?

I could imagine that in theory some users might want to do this:

```python
@task(
    container_image=ImageSpec(
        file="../Dockerfile.spark",
    )
)
def preprocess_spark():
    ...


@task(
    container_image=ImageSpec(
        # envd definition for a deepl-learning image based on torch
    )
)
def train_torch():
    ...
```

In this case it would be nice if both engines would be used for the respective task/image.
And even if only one image is used, the build engine could be already defined by the `ImageSpec`, right?

File: flytekit/image_spec/image_spec.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/image_spec/image_spec.py]: > python client picks up the creds from docker login.

yes, so people don't need to worry about it if they use docker hub.

we'll support passing docker credentials to pyflyte command. 
```
pyflyte run --docker_cred token workflow.py wf
```

[the remark this answers is not in the document]
Comment (fg91) [reply flytekit/image_spec/image_spec.py]: > yes, so people don't need to worry about it if they use docker hub.

For instance on GCP, the `gcloud` sdk [can act as a credentials helper](https://cloud.google.com/sdk/gcloud/reference/auth/configure-docker) for docker for images stored in google container registry or the newer artefact registry. But if I understand correctly, for e.g. a gcr.io image, flytekit would defer to the docker python sdk, this one to docker, docker to the registered credentials helper, done :)

**Comment by pingsutw** [comment]
Update:
ImageSpec can be used in two ways (CLI and task decorator). The image will be built before registration. if the image is found in the docker hub, flytekit will skip building the image.

1. CLI (pyflyte run / register / package / serialize)
```bash 
pyflyte run --image ImageSpec.yaml workflow.py wf
```
2. Task decorator
```python
@task(
    container_image=ImageSpec(packages=["colorama==0.2.1"], base_image="ghcr.io/flyteorg/flytekit:py3.10-1.4.2")
)
def train_torch():
    ...

```

**Review by wild-endeavor — APPROVED** [review]

**Comment by fg91** [comment]
> Update: ImageSpec can be used in two ways (CLI and task decorator). The image will be built before registration. if the image is found in the docker hub, flytekit will skip building the image.
> 
> 1. CLI (pyflyte run / register / package / serialize)
> 
> ```shell
> pyflyte run --image ImageSpec.yaml workflow.py wf
> ```
> 
> 2. Task decorator
> 
> ```python
> @task(
>     container_image=ImageSpec(packages=["colorama==0.2.1"], base_image="ghcr.io/flyteorg/flytekit:py3.10-1.4.2")
> )
> def train_torch():
>     ...
> ```

Very cool 🚀

**Comment by kumare3** [comment]
This is a just an amazing addition to Flyte - let's try it out for a bit and make it the default getting started

**Comment by fg91** [comment]
@ThomasWollmann fyi this feature now exists
